IdentityCore/src/validation/MSIDAuthority.m

```diff
@@ -137,4 +197,145 @@ + (NSURL *)cacheUrlForAuthority:(NSURL *)authority
     return [NSURL URLWithString:[NSString stringWithFormat:@"https://%@/%@", [authority msidHostWithPortIfNecessary], tenantId]];
 }
 
++ (void)discoverAuthority:(NSURL *)authority
+        userPrincipalName:(NSString *)upn
+                 validate:(BOOL)validate
+                  context:(id<MSIDRequestContext>)context
+          completionBlock:(MSIDAuthorityInfoBlock)completionBlock
+{
+    NSError *error;
+    if (![self isAuthorityFormatValid:authority context:nil error:&error])
+    {
+        if (completionBlock) { completionBlock(nil, NO, error); }
+        return;
+    }
+    
+    if (error)
+    {
+        if (completionBlock) completionBlock(nil, NO, error);
+        return;
+    }
+    
+    id <MSIDAuthorityResolverProtocol> resolver;
+    // ADFS.
+    if ([MSIDAuthority isADFSInstanceURL:authority])
+    {
+        resolver = [MSIDAdfsAuthorityResolver new];
+    }
+    // B2C.
+    else if ([MSIDAuthority isB2CInstanceURL:authority])
+    {
+        resolver = [MSIDB2CAuthorityResolver new];
+    }
+    // AAD.
+    else
+    {
+        resolver = [MSIDAadAuthorityResolver new];
+    }
+    
+    [resolver discoverAuthority:authority
+              userPrincipalName:upn
+                       validate:validate
+                        context:context
+                completionBlock:completionBlock];
+}
+
++ (void)loadOpenIdConfigurationInfo:(NSURL *)openIdConfigurationEndpoint
+                            context:(id<MSIDRequestContext>)context
+                    completionBlock:(MSIDOpenIdConfigurationInfoBlock)completionBlock
+{
+    __auto_type cacheKey = openIdConfigurationEndpoint.absoluteString.lowercaseString;
+    __auto_type metadata = [s_openIdConfigurationCache objectForKey:cacheKey];
+    
+    if (metadata)
+    {
+        if (completionBlock) completionBlock(metadata, nil);
+        return;
+    }
+    
+    __auto_type request = [[MSIDOpenIdConfigurationInfoRequest alloc] initWithEndpoint:openIdConfigurationEndpoint];
+    [request sendWithBlock:^(MSIDOpenIdProviderMetadata *metadata, NSError *error)
+    {
+        if (cacheKey && metadata)
+        {
+            [s_openIdConfigurationCache setObject:metadata forKey:cacheKey];
+        }
+        
+        if (completionBlock) completionBlock(metadata, error);
+    }];
+}
+
++ (NSURL *)normalizeAuthority:(NSURL *)authority
+                      context:(id<MSIDRequestContext>)context
+                        error:(NSError **)error
+{
+    if (![MSIDAuthority isAuthorityFormatValid:authority context:context error:error])
+    {
+        return nil;
+    }
+    
+    // B2C
+    if ([self isB2CInstanceURL:authority])
```

Comment:
consider adding a separate class for normalization and additional authority manipulations. e.g. separate for B2C, ADFS etc